functions/speech-to-speech/test/sample.integration.http.test.js

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright 2018, Google, LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-env node, mocha */
+
+const googleCloudProject = process.env.GOOGLE_CLOUD_PROJECT;
+const supertest = require('supertest');
+const request = supertest(process.env.BASE_URL);
+const fs = require('fs');
+const storageClient = getStorageClient();
+const storageLocation = 'US';
+const storageClass = 'MULTI_REGIONAL';
+const bucket = storageClient.bucket(process.env.OUTPUT_BUCKET);
+
+describe('Tests that do not require a GCS bucket', function () {
+  it('should fail if sampleRateHertz field is invalid.', function (done) {
+    request
+      .post('/speechTranslate')
+      .send({
+        encoding: 'LINEAR16',
+        sampleRateHertz: -24000,
+        languageCode: 'en',
+        audioContent: fs.readFileSync('test/speech-recording.b64', 'utf8')
+      })
+      .expect(400, /INVALID_ARGUMENT: sample_rate_hertz/, done);
+  });
+  it('should fail if languageCode field is invalid.', function (done) {
+    request
+      .post('/speechTranslate')
+      .send({
+        encoding: 'LINEAR16',
+        sampleRateHertz: 24000,
+        languageCode: 'xx',
+        audioContent: fs.readFileSync('test/speech-recording.b64', 'utf8')
+      })
+      .expect(400, /INVALID_ARGUMENT.*Invalid recognition.*bad language code/, done);
+  });
+  it('should fail if audioContent field is invalid.', function (done) {
+    request
+      .post('/speechTranslate')
+      .send({
+        encoding: 'LINEAR16',
+        sampleRateHertz: 24000,
+        languageCode: 'en',
+        audioContent: 'invalid-base64-audio-content'
+      })
+      .expect(400, 'invalid encoding', done);
+  });
+});
+
+describe('Tests that require a GCS bucket', function () {
+  // Create the Google Cloud Storage bucket to run the tests.
+  // Caution: The 'after' hook deletes the bucket. To prevent deleting user data, the 'before' hook
+  // fails if the bucket already exists.
+  let bucketCreated = false;
+  before('Create GCS bucket.', function (done) {
+    bucket.create({
```

Comment:
only create if `bucketCreated` is still false

It blows up for any tests where the bucket already exists

```
> speech-to-speech@0.0.1 system-test /tmpfs/src/github/nodejs-docs-samples/functions/speech-to-speech
> mocha --timeout 20000 test/sample.integration.http.test.js



  Tests that do not require a GCS bucket
    ✓ should fail if sampleRateHertz field is invalid. (217ms)
    ✓ should fail if languageCode field is invalid. (57ms)
    ✓ should fail if audioContent field is invalid.

  Tests that require a GCS bucket
    1) "before all" hook: Create GCS bucket.
    2) "after all" hook: Delete GCS bucket, files, and other test artifacts.


  3 passing (770ms)
  2 failing

  1) Tests that require a GCS bucket
       "before all" hook: Create GCS bucket.:
     Error: You already own this bucket. Please select another name.
      at ApiError (node_modules/@google-cloud/storage/node_modules/@google-cloud/common/build/src/util.js:43:9)
      at Util.parseHttpRespBody (node_modules/@google-cloud/storage/node_modules/@google-cloud/common/build/src/util.js:172:38)
      at Util.handleResp (node_modules/@google-cloud/storage/node_modules/@google-cloud/common/build/src/util.js:116:117)
      at retryRequest (node_modules/@google-cloud/storage/node_modules/@google-cloud/common/build/src/util.js:403:22)
      at onResponse (node_modules/retry-request/index.js:200:7)
      at /tmpfs/src/github/nodejs-docs-samples/functions/speech-to-speech/node_modules/teeny-request/build/src/index.js:191:13
      at process._tickCallback (internal/process/next_tick.js:109:7)

  2) Tests that require a GCS bucket
       "after all" hook: Delete GCS bucket, files, and other test artifacts.:
     Error: The before hook did not create the bucket, abort cleanup
      at Context.<anonymous> (test/sample.integration.http.test.js:105:12)  <-- <-- <--
```

Reply:
Oh nevermind, I didn't see that this describe block only has 1 test

So presumable the passed in OUTPUT_BUCKET already exists

Update code to find or create? Or TAL at what other tests which require a bucket are doing :)

Reply:
Mmm... changing the `before` hook to findOrCreate should be an easy fix.
However, to prevent accidentally deleting user files, I won't delete the bucket in the `after` hook. Let me know if that sounds okay to you.

Reply:
> "to prevent accidentally deleting user files"

Hmm. I hadn't considered that for running tests locally, good point!

I recommend having the tests create a unique bucket + run tests + delete the bucket <-- how most/many of our tests work

Take a look at this test for inspiration: [`recognize.test.js`](https://github.com/googleapis/nodejs-speech/blob/master/samples/system-test/recognize.test.js)

Relevant bits:

```js
const {Storage} = require(`@google-cloud/storage`);
const uuid = require(`uuid`);
const bucketName = `nodejs-docs-samples-test-${uuid.v4()}`;

test.before(async () => {
  const [bucket] = await storage.createBucket(bucketName);
  await bucket.upload(filepath);  // <-- we usually upload all of the required files
  await bucket.upload(filepath1); // <-- in the before() but you could do it in the test
});

test.after.always(async () => {
  const bucket = storage.bucket(bucketName);
  await bucket.deleteFiles({force: true});
  await bucket.deleteFiles({force: true}); // Try a second time...
  await bucket.delete();
});
```

Reply:
Mmm that sounds good to me, let me give it a shot.

Reply:
I tried the suggestion of having the test to create a bucket with a randomized name. However, because the Cloud Function uses env variables to get the bucket name I can't create the random name in a test. The reason is because the functions emulator doesn't take the new env variable until it restarts.

The alternative is to create the random name in the environment variable and then restart the emulator in the npm test script, as follows:

```
"test": "export OUTPUT_BUCKET=speech-to-speech-test-$(uuidgen) && functions restart && npm run local-test && npm run system-test"
```
I added this change in 99b6b0fae4e1fb266ba7404730842e1d1a584566. Let me know what you think.

Reply:
> because the Cloud Function uses env variables to get the bucket name

Ah, gotcha!

My recommendation is to go with this solution:

> Mmm... changing the before hook to findOrCreate should be an easy fix.
However, to prevent accidentally deleting user files, I won't delete the bucket in the after hook. Let me know if that sounds okay to you.

Change to findOrCreate & the tests should be green & I'm LGTM!

Reply:
Ok! Implemented findOrCreate and the tests do not remove the bucket. I tested locally with both scenarios:

1. The bucket exists previously.
1. The bucket doesn't exist.

See e9ef4ed8bd00d79bcee5bb5d2e4a9df6aaa9b2a8.